### PR TITLE
DSO: Fix the VMS DSO name converter to actually do something

### PR DESCRIPTION
**Pull request by levitte**
This function has never before actually done its work.  This wasn't
discovered before, because its output wasn't important before the FIPS
provider self test started using its value.

This function is now made to insert the VMS DSO extension (".EXE") at
the end of the filename, being careful to make sure what can be a
typical VMS generation number (separated from the file name with a
';') remains at the end.

### CODE REVIEW

File: crypto/dso/dso_vms.c [hunk 1]
```diff
@@ -453,11 +453,39 @@ static char *vms_merger(DSO *dso, const char *filespec1,
...
+        transform = 0;
+    } else {
+        p = filename;
+        transform = (strrchr(p, '>') == NULL || strrchr(p, ']') == NULL);
+    }
+
+    if (transform) {
+        int rsize = len + sizeof(DSO_EXTENSION);
+
+        translated = OPENSSL_malloc(rsize);
```
Comment (mattcaswell) [inline crypto/dso/dso_vms.c:473]: Should there be a malloc err raised if this fails?

File: crypto/dso/dso_vms.c [hunk 2]
```diff
@@ -453,11 +453,39 @@ static char *vms_merger(DSO *dso, const char *filespec1,
...
+            if (p != NULL)
+                len = p - filename;
+            strncpy(translated, filename, len);
+            translated[len] = '\0';
+            strcat(translated, DSO_EXTENSION);
+            if (p != NULL)
+                strcat(translated, p);
+        }
+    } else {
+        translated = OPENSSL_strdup(filename);
```
Comment (mattcaswell) [inline crypto/dso/dso_vms.c:486]: Similarly here

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: crypto/dso/dso_vms.c (hunk 1 above)
Comment (levitte) [reply crypto/dso/dso_vms.c:473]: None of the other DSO backends do that, and the API (in `dso_lib.c`) adds proper records, so I think not.

Comment (mattcaswell) [reply crypto/dso/dso_vms.c:473]: ok

**Review by mattcaswell — APPROVED** [review]

File: crypto/dso/dso_vms.c [hunk 3]
```diff
@@ -453,11 +453,39 @@ static char *vms_merger(DSO *dso, const char *filespec1,
...
+    const char *p;
+
+    len = strlen(filename);
+
+    p = strchr(filename, ':');
+    if (p != NULL) {
+        transform = 0;
+    } else {
+        p = filename;
+        transform = (strrchr(p, '>') == NULL || strrchr(p, ']') == NULL);
```
Comment (t8m) [inline crypto/dso/dso_vms.c:467]: This should be `&&` if I understand this right.

File: crypto/dso/dso_vms.c [hunk 4]
```diff
@@ -453,11 +453,39 @@ static char *vms_merger(DSO *dso, const char *filespec1,
...
+    }
+
+    if (transform) {
+        int rsize = len + sizeof(DSO_EXTENSION);
+
+        translated = OPENSSL_malloc(rsize);
+
```
Comment (t8m) [inline crypto/dso/dso_vms.c:474]: Maybe this empty line is not needed?

**Review by t8m — CHANGES_REQUESTED** [review]

File: crypto/dso/dso_vms.c (hunk 3 above)
Comment (levitte) [reply crypto/dso/dso_vms.c:467]: You understand very well.  Good catch!

File: crypto/dso/dso_vms.c (hunk 4 above)
Comment (levitte) [reply crypto/dso/dso_vms.c:474]: Sure

**Review by mattcaswell — APPROVED** [review]
Reconfirm

**Comment by mattcaswell** [comment]
Pushed.

**Pull request closed without merging**